Dispatcher: Get request URI from Symfony's Request

michaelKaefer opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Refactored method and mainly: get request URI from Symfony's request object. It's more reliable and future proof.
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | -
| How to test?  | -

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9065)
<!-- Reviewable:end -->

michaelKaefer committed:
Dispatcher: Get request URI from Symfony's Request

Refactored method and mainly: get request URI from Symfony's request object. It's more reliable and future proof. I really like the idea to "get rid of the home-made framework which is not our core business features" from http://build.prestashop.com/news/prestashop-1-7-and-symfony/. Hope this is ok. Are there plans to rebuild the Dispatcher class with Symfony components by the way?

mickaelandrieu commented on classes/Dispatcher.php:
```diff
@@ -437,35 +439,37 @@ public function dispatch()
+     *
+     * @param SymfonyRequest $symfonyRequest To retrieve the request URI from it
      */
-    protected function setRequestUri()
```
we can't change the definition and visibility of a function if it can be used/reused by another piece of code, this would force us to release PrestaShop 1.8 :trollface: 

This is what you can do when you want to improve a function:
* create your own - let's say `setRequestUriFromGolbals` (and you can inject Symfony Request as dependency, good idea!)
* use this function instead of the current one everywhere in PrestaShop code
* deprecate the current one for the latest release

This is boring work, but this ensure we don't break nothing for 1.7 users and developers :)

Let me know if you need more information, I agree for the changes you want to introduce and we will merge it if we can ;)

Good job!

michaelKaefer replied:
@mickaelandrieu Thanky you very much for your information! I will do it the next time :)

mickaelandrieu reviewed (changes requested):
minor changes to keep backward compatibility

michaelKaefer committed:
Use Symfony's request to build request URI

Leave it optional to inject a Request object (to not break existing code) to the dispatcher. If not injected create one and save it in a property. This way the Dispatcher has access to the Symfony Request object and can build the request URI from it (and is able to use it somewhere else too or pass it around)

michaelKaefer committed:
doc

michaelKaefer committed:
spaces

michaelKaefer commented:
@mickaelandrieu It took me quite some hours to think about about your points Thanks again! Is it better now? I took your approach but added some things. The old method is still used but as a wrapper around the new one which collects the dependencies (which now gets all dependencies injected and is easy to test for the most part). And the Request now is available to the whole dispatcher (which could be really cool for other things).

mickaelandrieu commented:
Thanks, it's good to me.

I have the feeling you really enjoy to contribute, if you want I can help/mentor you on something really great we are working on, like migrate pages from back office in Symfony or refactoring of some parts.

You'll learn a lot if you plan to make PrestaShop 1.7 websites/modules/themes.

Let me know if you're interested, we can talk about it when you're available for 👍

michaelKaefer commented:
@mickaelandrieu Thanks for merging! - And, yes, definitly, of course I am interested :+1: But how does it work? What can I do?

mickaelandrieu commented:
you can contact me using the gitter of PrestaShop in general channel and talk about what you may want to do https://gitter.im/PrestaShop/General

mickaelandrieu commented:
or by email: mickael [dot ] andrieu [at] prestashop [dot] com
